dvc/fs/azure.py

```diff
@@ -1,58 +1,94 @@
 from dvc.path_info import CloudURLInfo
 from dvc.progress import Tqdm
 from dvc.scheme import Schemes
+from dvc.utils import format_link
 
 from .base import BaseFileSystem
 
 logger = logging.getLogger(__name__)
+_DEFAULT_CREDS_STEPS = (
+    "https://azuresdkdocs.blob.core.windows.net/$web/python/"
+    "azure-identity/1.4.0/azure.identity.html#azure.identity"
+    ".DefaultAzureCredential"
+)
+
+
+class AzureAuthError(DvcException):
+    pass
 
 
 class AzureFileSystem(BaseFileSystem):
     scheme = Schemes.AZURE
     PATH_CLS = CloudURLInfo
+    PARAM_CHECKSUM = "etag"
+    DETAIL_FIELDS = frozenset(("etag", "size"))
     REQUIRES = {
-        "azure-storage-blob": "azure.storage.blob",
+        "adlfs": "adlfs",
         "knack": "knack",
+        "azure-identity": "azure.identity",
     }
-    PARAM_CHECKSUM = "etag"
-    DETAIL_FIELDS = frozenset(("etag", "size"))
-
-    COPY_POLL_SECONDS = 5
-    LIST_OBJECT_PAGE_SIZE = 5000
 
     def __init__(self, repo, config):
         super().__init__(repo, config)
 
-        url = config.get("url", "azure://")
+        url = config.get("url")
         self.path_info = self.PATH_CLS(url)
         self.bucket = self.path_info.bucket
 
         if not self.bucket:
             container = self._az_config.get("storage", "container_name", None)
-            self.path_info = self.PATH_CLS(f"azure://{container}")
-            self.bucket = self.path_info.bucket
+            url = f"azure://{container}"
 
-        self._conn_str = config.get(
-            "connection_string"
-        ) or self._az_config.get("storage", "connection_string", None)
+        self.path_info = self.PATH_CLS(url)
+        self.bucket = self.path_info.bucket
+
+        self.login_method, self.login_info = self._prepare_credentials(config)
 
-        self._account_url = None
-        if not self._conn_str:
-            name = self._az_config.get("storage", "account", None)
-            self._account_url = f"https://{name}.blob.core.windows.net"
+    def _prepare_credentials(self, config):
+        from azure.identity.aio import DefaultAzureCredential
 
-        self._credential = config.get("sas_token") or self._az_config.get(
-            "storage", "sas_token", None
+        login_info = {}
+        login_info["connection_string"] = config.get(
+            "connection_string",
+            self._az_config.get("storage", "connection_string", None),
+        )
+        login_info["account_name"] = config.get(
+            "account_name", self._az_config.get("storage", "account", None)
+        )
+        login_info["account_key"] = config.get(
+            "account_key", self._az_config.get("storage", "key", None)
+        )
+        login_info["sas_token"] = config.get(
+            "sas_token", self._az_config.get("storage", "sas_token", None)
         )
-        if not self._credential:
-            self._credential = self._az_config.get("storage", "key", None)
+        login_info["tenant_id"] = config.get("tenant_id")
```

Comment:
can they also com from storage?

Reply:
No, at least it isn't listed here https://docs.microsoft.com/en-us/cli/azure/azure-cli-configuration#cli-configuration-values-and-environment-variables

Reply:
gotcha, thanks!